### PR TITLE
Remove "bower.json" and only create if necessary

### PR DESCRIPTION
**Pull request by Turbo87**
"if necessary" being if a blueprint specifically wants to install Bower dependencies.

This PR removes the need for Ember CLI to lazily install Bower during `ember new`, since the `bower.json` file is now gone in the `app` and `addon` blueprints.

### CODE REVIEW

File: lib/tasks/bower-install.js [hunk 1]
```diff
@@ -1,32 +1,48 @@
...
+    return this.resolveBower()
       .catch(() => this.installBower()
-        .then(() => resolve('bower', { basedir: cliPath })))
-      .then(bowerPath => require(bowerPath));
+        .then(() => this.resolveBower()))
+      .then(bowerPath => this.importBower(bowerPath));
   }
 
   installBower() {
+    logger.info('Installing "bower" via NPM into: %s', cliPath);
```
Comment (twokul) [inline lib/tasks/bower-install.js]: what's `cliPath` here? global or `node_modules/ember-cli`?

File: lib/tasks/bower-install.js [hunk 2]
```diff
@@ -37,32 +53,61 @@ class BowerInstallTask extends Task {
...
 
-    if (!existsSync(bowerJson)) {
-      ui.writeWarnLine('Skipping bower install: bower.json not found');
+    let packages = options.packages || [];
+    let installOptions = options.installOptions || {};
+    let savePackages = installOptions.save || installOptions.saveDev;
+
+    // if we are running "bower install" from "ember init" and there is
+    // no "bower.json" we return early
+    if (!this.hasBowerJson() && !savePackages) {
```
Comment (twokul) [inline lib/tasks/bower-install.js]: is `!savePackages` necessary here?

File: tests/unit/tasks/bower-install-test.js [hunk 3]
```diff
@@ -2,31 +2,60 @@
...
+      });
+    });
+
+    it('install "bower" if it can not be resolved', function() {
+      td.when(task.resolveBower()).thenReturn(Promise.reject(), Promise.resolve('path/to/bower'));
+      td.when(task.installBower()).thenResolve();
+      td.when(task.importBower('path/to/bower')).thenReturn('ok');
+
+      return task.ensureBower().then(ok => {
+        expect(ok).to.equal('ok');
```
Comment (twokul) [inline tests/unit/tasks/bower-install-test.js]: 👌 is👌

**Comment by twokul** [comment]
restarting Appveyor

File: lib/tasks/bower-install.js (hunk 1 above)
Comment (Turbo87) [inline lib/tasks/bower-install.js]: it depends ™

if you're in `ember new` it will be the global, if you're in `ember install` it will be the local

File: lib/tasks/bower-install.js (hunk 2 above)
Comment (Turbo87) [inline lib/tasks/bower-install.js]: yes, otherwise `addBowerPackagesToProject()` won't actually add anything to `bower.json` if it doesn't exist yet

File: tests/unit/tasks/bower-install-test.js (hunk 3 above)
Comment (Turbo87) [inline tests/unit/tasks/bower-install-test.js]: 🤷‍♂️

**Comment by Turbo87** [comment]
I seriously have no idea anymore what's wrong with AppVeyor. The code itself seems to work fine, but CI is :trollface:ing us...

File: lib/tasks/bower-install.js [hunk 4]
```diff
@@ -37,32 +53,61 @@ class BowerInstallTask extends Task {
...
+      logger.info('Skipping "bower install" since "bower.json" does not exist');
       return Promise.resolve();
     }
 
-    return this.ensureBower().then(bower => {
+    // resolve "bower" and install via NPM if necessary
+    // -> then check if a "bower.json" file exists and create if necessary
+    // -> then run the actual "bower install"
+    return this.ensureBower().then(() => this.ensureBowerJson()).then(() => {
```
Comment (nathanhammond) [inline lib/tasks/bower-install.js]:
```javascript
RSVP.all([this.ensureBower(), this.ensureBowerJson()])
  .then(...)
```

File: lib/tasks/bower-install.js (hunk 1 above)
Comment (nathanhammond) [inline lib/tasks/bower-install.js]: This somewhat falls in line with our goals of using global package managers. Note that we don't trigger `--save`. @Turbo87 I'd prefer to _only_ do `npm install -g bower`...

File: blueprints/app/files/bower.json [hunk 5]
```diff
@@ -1,5 +0,0 @@
-{
```
Comment (nathanhammond) [inline blueprints/app/files/bower.json]: I'm tempted to move this into a different generator rather than just omitting it. `bower init` is _way_ overkill for something that will never be run as a `bower` package.

```
$ bower install --save 'jquery'
bower cached        https://github.com/jquery/jquery-dist.git#3.1.1
bower validate      3.1.1 against https://github.com/jquery/jquery-dist.git#*
bower no-json       No bower.json file to save to, use bower init to create one
```

(For people manually installing `bower` packages instead of transitively via `ember install`.)

File: lib/models/blueprint.js [hunk 6]
```diff
@@ -1017,7 +1017,7 @@ let Blueprint = CoreObject.extend({
     return task.run({
       verbose: true,
       packages: packageNamesAndVersions,
-      installOptions,
+      installOptions: installOptions || { save: true },
```
Comment (nathanhammond) [inline lib/models/blueprint.js]: This should be `merge`d. (I'm aware that the original didn't merge and that this is simply moved.)

File: lib/tasks/bower-install.js [hunk 7]
```diff
@@ -37,32 +53,61 @@ class BowerInstallTask extends Task {
...
+  hasBowerJson() {
+    return existsSync(this.bowerJsonPath);
+  }
+
+  ensureBowerJson() {
+    if (this.hasBowerJson()) {
+      return Promise.resolve();
+    }
+
+    let projectName = this.project.name();
```
Comment (nathanhammond) [inline lib/tasks/bower-install.js]: - ✅ `in-repo-addon`
- ✅ `app`
- ✅ `addon` (since it will be the dummy app)

Also, not actually used or relevant.

File: lib/tasks/bower-install.js (hunk 2 above)
Comment (nathanhammond) [inline lib/tasks/bower-install.js]: Note: rather than omitting the file we can do `Object.keys().length` of `dependencies` and `devDependencies` to determine if we need to install it. It's also less likely to break `ember-try`.

File: blueprints/app/files/bower.json (hunk 5 above)
Comment (nathanhammond) [inline blueprints/app/files/bower.json]: (This comment only applies if we maintain the same approach of `hasBowerJson`.)

File: lib/tasks/bower-install.js (hunk 1 above)
Comment (Turbo87) [inline lib/tasks/bower-install.js]: @nathanhammond we have no control over the installed bower version in that case

File: lib/tasks/bower-install.js (hunk 2 above)
Comment (Turbo87) [inline lib/tasks/bower-install.js]: we could, but then again, why have an essentially empty file in the project?

File: lib/tasks/bower-install.js (hunk 4 above)
Comment (Turbo87) [inline lib/tasks/bower-install.js]: in that case you might get a `bower.json` file without the dependency if the `npm install bower` fails

File: blueprints/app/files/bower.json (hunk 5 above)
Comment (Turbo87) [inline blueprints/app/files/bower.json]: not sure how `bower init` is involved here

File: lib/models/blueprint.js (hunk 6 above)
Comment (Turbo87) [inline lib/models/blueprint.js]: that would technically be a breaking API change

Comment (nathanhammond) [inline lib/models/blueprint.js]: I'd call it a bugfix. 😜 YOLO?

File: blueprints/app/files/bower.json (hunk 5 above)
Comment (nathanhammond) [inline blueprints/app/files/bower.json]: This is an ergonomics question, paired with https://github.com/ember-cli/ember-cli/pull/6698#discussion_r96703772 ... a blank `bower.json` file means that users can blissfully use `bower`. Otherwise the user is presented with the above message asking them to `bower init` which is _way_ more than they need.

File: lib/tasks/bower-install.js (hunk 4 above)
Comment (nathanhammond) [inline lib/tasks/bower-install.js]: Don't care too strongly. Can you promote the `.then` chain to the next line for clarity?

File: lib/tasks/bower-install.js (hunk 1 above)
Comment (nathanhammond) [inline lib/tasks/bower-install.js]: @Turbo87 We can

1. Check to see if there is a global, and if it is above our version. If so, we use it.
2. If no global, install one, return.
3. If global and under our version, install into the current `cliPath`.

Note: I'd prefer to always dump this into our top-level `ember-cli` instead of into the project's copy. You'll counter that version matters local to the version of CLI and I'd agree, especially for `bower` since we interact with it programmatically instead of shelling out, and we'll make no changes.

Comment (Turbo87) [inline lib/tasks/bower-install.js]: There are several issues with what you're proposing since we're not shelling out to bower but requiring it instead. Also using global bower seems unrelated to this PR and can be discussed separately.

File: blueprints/app/files/bower.json (hunk 5 above)
Comment (Turbo87) [inline blueprints/app/files/bower.json]: The user will not be presented with any message!? We automatically create the file if we have to, though not through bower init.

File: lib/tasks/bower-install.js (hunk 1 above)
Comment (nathanhammond) [inline lib/tasks/bower-install.js]: This conversation went exactly as I expected, I'm comfortable with things as they stand. 😜 

> You'll counter that version matters local to the version of CLI and I'd agree, especially for bower since we interact with it programmatically instead of shelling out, and we'll make no changes.

**Comment by Turbo87** [comment]
@stefanpenner @rwjblue friendly reminder 😉

File: blueprints/app/files/bower.json (hunk 5 above)
Comment (stefanpenner) [inline blueprints/app/files/bower.json]: let me know if i misunderstood this comment thread, but it sounds like:

* @nathanhammond is saying: using a warning a user would get when using `bower` cli directly themselves
*  @Turbo87 is saying: when ember cli invokes `bower`, it handles this case already.

If so, it seems reasonable to me that a user intending to once again use bower, is warned to first use `bower init`.

File: lib/models/blueprint.js (hunk 6 above)
Comment (stefanpenner) [inline lib/models/blueprint.js]: Sounds reasonable as a `[BUGFIX]` commit, independent of this PR.

File: lib/tasks/bower-install.js [hunk 8]
```diff
@@ -1,32 +1,48 @@
...
+
   ensureBower() {
     if (this.bower) {
-      return Promise.resolve(this.bower);
+      return Promise.resolve();
     }
 
-    return resolve('bower', { basedir: cliPath })
+    return this.resolveBower()
       .catch(() => this.installBower()
```
Comment (stefanpenner) [inline lib/tasks/bower-install.js]: I would prefer if we only installed bower if a specific error occurs. For example, I would hate for some other issue (SyntaxError or something), from accidentally causing `bower` to be installed

File: lib/tasks/bower-install.js (hunk 7 above)
Comment (stefanpenner) [inline lib/tasks/bower-install.js]: vanity nit-pick: if this was `let name = this.project.name()` the `writeJson` bellow could use concise object syntax `{ name }`

**Comment by stefanpenner** [comment]
with https://github.com/ember-cli/ember-cli/pull/6698#discussion_r97881345 fixed, you have my 👍  The other is merely a nit-pick.

**Comment by Turbo87** [comment]
- AppVeyor is broken as always
- @nathanhammond seems to agree with the PR now
- addressed @stefanpenner's comments

anyone please confirm and press the big (currently not so green) merge button 😉 

(do **not** use @homu to merge this, as I've changed the base branch for this PR)

File: lib/tasks/bower-install.js [hunk 9]
```diff
@@ -1,32 +1,53 @@
...
+      return Promise.resolve();
     }
 
-    return resolve('bower', { basedir: cliPath })
-      .catch(() => this.installBower()
-        .then(() => resolve('bower', { basedir: cliPath })))
-      .then(bowerPath => require(bowerPath));
+    return this.resolveBower()
+      .catch(error => {
+        if (error.message.indexOf('Cannot find module \'bower\'') === -1) {
```
Comment (rwjblue) [inline lib/tasks/bower-install.js]: Is the error message platform dependent or guaranteed?

Comment (Turbo87) [inline lib/tasks/bower-install.js]: tl;dr this is the best we can do for now... (see https://github.com/substack/node-resolve/commit/c622aefeb286e479d536601e30bb828e69f86ec3#commitcomment-20625328)

**Comment by rwjblue** [comment]
👏
